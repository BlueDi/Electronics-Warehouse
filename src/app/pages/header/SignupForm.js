import React, { Component } from 'react';
import { withCookies } from 'react-cookie';
import { withRouter } from 'react-router-dom';
import { Form, Header, Image, Message } from 'semantic-ui-react';
import { service } from '@utils';
import logo from '@assets/images/logo.png';

class SignupForm extends Component {
  state = {
    error: false,
    error_header: '',
    error_detail: '',
    name: '',
    password: '',
    email: '',
    permission: ''
  };

  handleChange = (e, { name, value }) =>
    this.setState({ error: false, [name]: value });

  handleValidation = () => {
    const body = {
      name: this.state.name,
      password: this.state.password,
      email: this.state.email,
      permission: this.state.permission
    };
    service
      .post('/signup', body)
      .then(response => {
        this.setupLoggedUser(response.data);
      })
      .catch(err => {
        this.failedSignup(err.response.data);
      });
  };

  setupLoggedUser = value => {
    const { cookies } = this.props;
    for (var property in value) {
      cookies.set(property, value[property], { path: '/' });
    }
    cookies.set('cart', [], { path: '/' });
  };

  failedSignup = err => {
    this.setState({
      error: true,
      error_header: err.header,
      error_detail: err.detail
    });
  };

  nameField = () => {
    return (
      <Form.Input
        fluid
        icon="user"
        iconPosition="left"
        label="Name"
        name="name"
        placeholder="User name"
        required
        onChange={this.handleChange}
      />
    );
  };

  emailField = () => {
    return (
      <Form.Input
        fluid
        icon="at"
        iconPosition="left"
        name="email"
        label="E-mail"
        type="email"
        placeholder="E-mail address"
        required
        onChange={this.handleChange}
      />
    );
  };

  passwordField = () => {
    return (
      <Form.Input
        fluid
        icon="lock"
        iconPosition="left"
        name="password"
        label="Password"
        placeholder="Password"
        required
        type="password"
        onChange={this.handleChange}
      />
    );
  };

  permissionField = () => {
    return (
      <Form.Input
        fluid
        icon="options"
        iconPosition="left"
        name="permission"
        label="Permission Key"
        type="password"
        placeholder="Permission key"
        onChange={this.handleChange}
      />
    );
  };

  errorMessage = () => {
    return (
      <Message
        error
        header={this.state.error_header}
        content={this.state.error_detail}
      />
    );
  };

  render() {
    return (
      <React.Fragment>
        <Header size="large">
          <Image src={logo} /> Signup an account
        </Header>
        <Form
          size="large"
          error={this.state.error}
          onSubmit={this.handleValidation}
        >
          {this.nameField()}
          {this.emailField()}
          {this.passwordField()}
          {this.permissionField()}
          {this.errorMessage()}
          <Form.Button fluid>Signup</Form.Button>
        </Form>
      </React.Fragment>
    );
  }
}

export default withRouter(withCookies(SignupForm));
