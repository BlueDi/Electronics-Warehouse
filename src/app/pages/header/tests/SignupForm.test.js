import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter as Router } from 'react-router-dom';
import SignupForm from '@pages/header/SignupForm';

describe('<SignupForm />', () => {
  const component = renderer.create(
    <Router>
      <SignupForm />
    </Router>
  );

  test('Renders properly', () => {
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Renders error message properly', () => {
    const instance = component.root;
    const input_name = instance.find(
      el => el.type === 'input' && el.props.name === 'name'
    );
    input_name.props.onChange({
      target: { value: 'stu' }
    });

    const form = instance.find(el => el.type === 'form');
    form.props.onSubmit();

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
