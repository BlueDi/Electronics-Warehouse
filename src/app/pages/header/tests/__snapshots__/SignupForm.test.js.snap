// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignupForm /> Renders error message properly 1`] = `
Array [
  <div
    className="ui large header"
  >
    <img
      className="ui image"
      src="test-file-stub"
    />
     Signup an account
  </div>,
  <form
    className="ui large form"
    onSubmit={[Function]}
  >
    <div
      className="required field"
    >
      <label>
        Name
      </label>
      <div
        className="ui fluid left icon input"
      >
        <input
          name="name"
          onChange={[Function]}
          placeholder="User name"
          required={true}
          type="text"
        />
        <i
          aria-hidden="true"
          className="user icon"
        />
      </div>
    </div>
    <div
      className="required field"
    >
      <label>
        E-mail
      </label>
      <div
        className="ui fluid left icon input"
      >
        <input
          name="email"
          onChange={[Function]}
          placeholder="E-mail address"
          required={true}
          type="email"
        />
        <i
          aria-hidden="true"
          className="at icon"
        />
      </div>
    </div>
    <div
      className="required field"
    >
      <label>
        Password
      </label>
      <div
        className="ui fluid left icon input"
      >
        <input
          name="password"
          onChange={[Function]}
          placeholder="Password"
          required={true}
          type="password"
        />
        <i
          aria-hidden="true"
          className="lock icon"
        />
      </div>
    </div>
    <div
      className="field"
    >
      <label>
        Permission Key
      </label>
      <div
        className="ui fluid left icon input"
      >
        <input
          name="permission"
          onChange={[Function]}
          placeholder="Permission key"
          type="password"
        />
        <i
          aria-hidden="true"
          className="options icon"
        />
      </div>
    </div>
    <div
      className="ui error message"
    >
      <div
        className="content"
      >
        <div
          className="header"
        >
          
        </div>
        <p>
          
        </p>
      </div>
    </div>
    <div
      className="field"
    >
      <button
        className="ui fluid button"
        onClick={[Function]}
        role="button"
      >
        Signup
      </button>
    </div>
  </form>,
]
`;

exports[`<SignupForm /> Renders properly 1`] = `
Array [
  <div
    className="ui large header"
  >
    <img
      className="ui image"
      src="test-file-stub"
    />
     Signup an account
  </div>,
  <form
    className="ui large form"
    onSubmit={[Function]}
  >
    <div
      className="required field"
    >
      <label>
        Name
      </label>
      <div
        className="ui fluid left icon input"
      >
        <input
          name="name"
          onChange={[Function]}
          placeholder="User name"
          required={true}
          type="text"
        />
        <i
          aria-hidden="true"
          className="user icon"
        />
      </div>
    </div>
    <div
      className="required field"
    >
      <label>
        E-mail
      </label>
      <div
        className="ui fluid left icon input"
      >
        <input
          name="email"
          onChange={[Function]}
          placeholder="E-mail address"
          required={true}
          type="email"
        />
        <i
          aria-hidden="true"
          className="at icon"
        />
      </div>
    </div>
    <div
      className="required field"
    >
      <label>
        Password
      </label>
      <div
        className="ui fluid left icon input"
      >
        <input
          name="password"
          onChange={[Function]}
          placeholder="Password"
          required={true}
          type="password"
        />
        <i
          aria-hidden="true"
          className="lock icon"
        />
      </div>
    </div>
    <div
      className="field"
    >
      <label>
        Permission Key
      </label>
      <div
        className="ui fluid left icon input"
      >
        <input
          name="permission"
          onChange={[Function]}
          placeholder="Permission key"
          type="password"
        />
        <i
          aria-hidden="true"
          className="options icon"
        />
      </div>
    </div>
    <div
      className="ui error message"
    >
      <div
        className="content"
      >
        <div
          className="header"
        >
          
        </div>
        <p>
          
        </p>
      </div>
    </div>
    <div
      className="field"
    >
      <button
        className="ui fluid button"
        onClick={[Function]}
        role="button"
      >
        Signup
      </button>
    </div>
  </form>,
]
`;
